scripts/prune_lm_dir.py

```diff
@@ -328,6 +351,160 @@ def FinalizeOutput(final_work_out):
+        scale = threshold / args.threshold
+        steps += 'prune*{0} EM EM EM'.format(scale).split()
+
+    # Prune step
+    logprob_changes.append(RunStep(step, in_step=recovery_step))
+    step += 1
+    while step < len(steps): # EM steps
+      logprob_changes.append(RunStep(step))
+      step += 1
+    iter += 1
+
+    if MatchTargetSize(final_num_ngrams):
+        return (True, step, iter, recovery_step, recovery_size)
+
+    if iter > args.max_iter:
+        sys.exit("prune_lm_dir.py: Too many iterations, please set a higher threshold and rerun.")
+
+    # always prune from LM larger than target_num_ngrams
+    if final_num_ngrams > args.target_num_ngrams and (recovery_size <= 0 or final_num_ngrams < recovery_size):
+        recovery_step = step
+        recovery_size = final_num_ngrams
+
+    return (False, step, iter, recovery_step, recovery_size)
+
+class LinearEstimator(object):
+    """Estimate the coeffients of a line using two most recent points
+
+    Example Usage:
+
+        ls = LinearEstimator()
+        ls.AddPoint(x0, y0)
+        ls.AddPoint(x1, y1)
+
+        while True:
+            xn = Update()
+            yn = ls.Estimate(xn)
+
+            if Match(yn):
+                break
+
+            ls.AddPoint(xn, yn)
+    """
+    def __init__(self):
+        self.switch = True
+        self.x0 = 0.0
+        self.x1 = 0.0
+        self.y0 = 0.0
+        self.y1 = 0.0
+        self.n = 0
+
+    def AddPoint(self, x, y):
+        if self.switch:
+            self.x0 = x
+            self.y0 = y
+        else:
+            self.x1 = x
+            self.y1 = y
+        self.switch = not self.switch
+        self.n += 1
+
+    def Estimate(self, x):
+        assert self.n >= 2
+        alpha = (self.y1 - self.y0) / (self.x1 - self.x0)
+        beta = self.y0 - alpha * self.x0
+
+        return alpha * x + beta
+
+# find threshold in order to match the target-num-ngrams with final LM
+#
+# Here we fit a linear of log(num-ngrams) versus log(threshold)
+# with the latest two points, and approach the target-num-ngrams gradually.
+def FindThreshold():
+    global final_num_ngrams
+
+    step = 0
+    recovery_step = step
+    recovery_size = 0
+    iter = 0
+    threshold = args.threshold
+
+    ls = LinearEstimator()
+
+    # get initial two points
+    (success, step, iter, recovery_step, recovery_size) = \
+        IterateOnce(0.0, step, iter, recovery_step, recovery_size)
+    if success:
+        return (threshold, iter)
+
+    if final_num_ngrams < args.target_num_ngrams:
+        sys.exit("prune_lm_dir.py: --initial-threshold={0} is too big, "
+                 "please reduce this value and rerun. "
+                 "Number of n-grams pruning with this threshold is {1} "
+                 "versus --target-num-ngrams={2}".format(args.threshold,
+                     final_num_ngrams, args.target_num_ngrams))
+
+    ls.AddPoint(math.log(final_num_ngrams), math.log(threshold))
+
+    threshold = 1.5 * threshold
```

Comment:
I would change the statement 'threshold = 1.5 * threshold' to something like the following (and bear in mind I'm assuming you renamed final_num_ngrams -> current_num_ngrams).
```  
   # To get an initial estimate of the log-log linear regression between num-ngrams and
   # pruning threshold, we need at least two data points.  The initial data point is
   # obtained by pruning to the user-specified threshold 'args.initial_threshold'.
   # We get the next data point by multiplying 'threshold' by a value that can be
   # as much as 4.0, but will be less than that if we're relatively close to the
   # final desired model size (because we don't want to overshoot).
   threshold_increase_ratio = min(4.0,
                                                       (current_num_ngrams / final_num_ngrams) ** 0.333)
   threshold = threshold * threshold_increase_ratio
```
Also, please rename 'threshold' to 'cur_threshold', it makes it clearer that it's a varying quantity.
if current_num_grams > 4.0 * final_num_ngrams:
      threshold = 2.0 * threshold